Add facet_col and animation_frame argument to imshow (#2746)
plotly/plotly.py: doc/python/imshow.md

fig.show()
```

### Exploring 3-D images and timeseries with `facet_col`

*Introduced in plotly 4.13*

For three-dimensional image datasets, obtained for example by MRI or CT in medical imaging, one can explore the dataset by representing its different planes as facets. The `facet_col` argument specifies along which axis the image is sliced through to make the facets. With `facet_col_wrap`, one can set the maximum number of columns. For image datasets passed as xarrays, it is also possible to specify the axis by its name (label), thus passing a string to `facet_col`.

It is recommended to use `binary_string=True` for facetted plots of images in order to keep a small figure size and a short rendering time.

See the [tutorial on facet plots](/python/facet-plots/) for more information on creating and styling facet plots.

```python
import plotly.express as px

> from skimage import io
> from skimage.data import image_fetcher
> path = image_fetcher.fetch('data/cells.tif')
> data = io.imread(path)

Reviewer: ```suggestion
from skimage import util
from skimage.data import cells3d
data = util.img_as_float(cells3d()[:, 1, :, :])  # grab only the nuclei
```

Author: Hum these changes have not been released in scikit-image, have they?

Reviewer: You are right! The dev version tends to be *my* reference version, hence the confusion. Sorry, please disregard these suggestions. I'll delete them if it's straightforward to do so...

img = data[20:45:2]
fig = px.imshow(img, facet_col=0, binary_string=True, facet_col_wrap=5)
fig.show()
```

### Exploring 3-D images and timeseries with `animation_frame`

*Introduced in plotly 4.13*

For three-dimensional image datasets, obtained for example by MRI or CT in medical imaging, one can explore the dataset by sliding through its different planes in an animation. The `animation_frame` argument of `px.imshow` sets the axis along which the 3-D image is sliced in the animation.

```python
import plotly.express as px

> from skimage import io
> from skimage.data import image_fetcher
> path = image_fetcher.fetch('data/cells.tif')
> data = io.imread(path)

Reviewer: ```suggestion
from skimage import util
from skimage.data import cells3d
data = util.img_as_float(cells3d()[:, 1, :, :])  # grab only the nuclei
```

img = data[25:40]
fig = px.imshow(img, animation_frame=0, binary_string=True)
fig.show()
```

### Animations of xarray datasets

> *Introduced in plotly 4.11*

Author: ```suggestion
*Introduced in plotly 4.13*
```

For xarray datasets, one can pass either an axis number or an axis name to `animation_frame`. Axis names and coordinates are automatically used for the labels, ticks and animation controls of the figure.

```python
import plotly.express as px
import xarray as xr
# Load xarray from dataset included in the xarray tutorial
ds = xr.tutorial.open_dataset('air_temperature').air[:20]
fig = px.imshow(ds, animation_frame='time', zmin=220, zmax=300, color_continuous_scale='RdBu_r')
fig.show()
```

### Combining animations and facets

> It is possible to view 4-dimensional datasets (for example, 3-D images evolving with time) using a combination of `animation_frame` and `facet_col`.

Reviewer: ```suggestion
It is possible to view four-dimensional datasets (for example, 3-D images evolving with time) using a combination of `animation_frame` and `facet_col`.
```
(because above text reads 'three-dimensional')

```python
import plotly.express as px

> from skimage import io
> from skimage.data import image_fetcher
> path = image_fetcher.fetch('data/cells.tif')
> data = io.imread(path)

Reviewer: ```suggestion
from skimage import util
from skimage.data import cells3d
data = util.img_as_float(cells3d()[:, 1, :, :])  # grab only the nuclei
```

fig.show()
```

### Exploring 3-D images, timeseries and sequences of images with `facet_col`

*Introduced in plotly 4.14*

For three-dimensional image datasets, obtained for example by MRI or CT in medical imaging, one can explore the dataset by representing its different planes as facets. The `facet_col` argument specifies along which axis the image is sliced through to make the facets. With `facet_col_wrap`, one can set the maximum number of columns. For image datasets passed as xarrays, it is also possible to specify the axis by its name (label), thus passing a string to `facet_col`.

It is recommended to use `binary_string=True` for facetted plots of images in order to keep a small figure size and a short rendering time.

See the [tutorial on facet plots](/python/facet-plots/) for more information on creating and styling facet plots.

```python
import plotly.express as px
from skimage import io
from skimage.data import image_fetcher
path = image_fetcher.fetch('data/cells.tif')
data = io.imread(path)
img = data[20:45:2]
fig = px.imshow(img, facet_col=0, binary_string=True, facet_col_wrap=5)
fig.show()
```

> Facets can also be used to represent several images of equal shape, like in the example below where different values of the blurring parameter of a Gaussian filter are compared.

Author: I added this example @nicolaskruchten to show that sequences of images can also be represented with facets. Should we accept lists and call `np.array` on lists or tuples at the beginning of `px.imshow`? I have mixed feelings about this, because not any list will work (in particular arrays of different shapes) so if there is an error to be thrown I'd rather have it thrown in the user's code than by `imshow`.

Reviewer 2: let's let the user deal with it.

Author: what do you think about how facet titles are set in this example? Should we introduce a parameter for facet titles in imshow?

Reviewer 2: `labels=dict(facet_col="...")` is the general PX pattern here, yes, if you pass in an array to `facet_col=array`, and it shows up as `facet_col=<value>`. I would like to keep the idea of "if you see `x=` you can relabel it with `labels=dict(x="new")` which isn't the case with the default `"slice"` right now.

fig.show()
```

### Exploring 3-D images and timeseries with `facet_col`

*Introduced in plotly 4.11*

> For three-dimensional image datasets, obtained for example by MRI or CT in medical imaging, one can explore the dataset by representing its different planes as facets. The `facet_col` argument specifies along which axes the image is sliced through to make the facets. With `facet_col_wrap` , one can set the maximum number of columns. For image datasets passed as xarrays, it is also possible to give an axis name as a string for `facet_col`.

Reviewer: ```suggestion
For three-dimensional image datasets, obtained for example by MRI or CT in medical imaging, one can explore the dataset by representing its different planes as facets. The `facet_col` argument specifies along which axis the image is sliced through to make the facets. With `facet_col_wrap`, one can set the maximum number of columns. For image datasets passed as xarrays, it is also possible to specify the axis by its name (label), thus passing a string to `facet_col`.
```
